tsdb/engine/bz1/bz1.go

```diff
@@ -0,0 +1,601 @@
+	// Ignore if there are no fields to save.
+	if len(m) == 0 {
+		return nil
+	}
+
+	// Persist each measurement field in the map.
+	b := tx.Bucket([]byte("fields"))
+	for k, f := range m {
+		// Marshal field into bytes.
+		data, err := f.MarshalBinary()
+		if err != nil {
+			return fmt.Errorf("marshal measurement field: %s", err)
+		}
+
+		// Insert marshaled data into key.
+		if err := b.Put([]byte(k), data); err != nil {
+			return fmt.Errorf("put: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// WriteIndex writes marshaled points to the engine's underlying index.
+func (e *Engine) WriteIndex(pointsByKey map[string][][]byte) error {
+	return e.db.Update(func(tx *bolt.Tx) error {
+		for key, values := range pointsByKey {
+			if err := e.writeIndex(tx, key, values); err != nil {
+				return fmt.Errorf("write: key=%x, err=%s", key, err)
+			}
+		}
+		return nil
+	})
+}
+
+// writeIndex writes a set of points for a single key.
+func (e *Engine) writeIndex(tx *bolt.Tx, key string, a [][]byte) error {
+	// Ignore if there are no points.
+	if len(a) == 0 {
+		return nil
+	}
+
+	// Create or retrieve series bucket.
+	bkt, err := tx.Bucket([]byte("points")).CreateBucketIfNotExists([]byte(key))
+	if err != nil {
+		return fmt.Errorf("create series bucket: %s", err)
+	}
+	c := bkt.Cursor()
+
+	// Ensure the slice is sorted before retrieving the time range.
+	sort.Sort(byteSlices(a))
+
+	// Determine time range of new data.
+	tmin, tmax := int64(btou64(a[0][0:8])), int64(btou64(a[len(a)-1][0:8]))
+
+	// If tmin is after the last block then append new blocks.
+	//
+	// This is the optimized fast path. Otherwise we need to merge the points
+	// with existing blocks on disk and rewrite all the blocks for that range.
+	if k, v := c.Last(); k == nil || int64(btou64(v[0:8])) < tmin {
+		if err := e.writeBlocks(bkt, a); err != nil {
+			return fmt.Errorf("append blocks: %s", err)
+		}
+	}
+
+	// Generate map of inserted keys.
+	m := make(map[int64]struct{})
+	for _, b := range a {
+		m[int64(btou64(b[0:8]))] = struct{}{}
+	}
+
+	// If time range overlaps existing blocks then unpack full range and reinsert.
+	for k, v := c.First(); k != nil; k, v = c.Next() {
+		// Determine block range.
+		bmin, bmax := int64(btou64(k)), int64(btou64(v[0:8]))
+
+		// Exit loop if the lowest block time is greater than the max insert time.
+		if bmin > tmax {
+			break
+		}
+
+		// If range overlaps with inserted range then read out data and delete block.
+		if (bmin >= tmin && bmin <= tmax) || (bmax >= tmin && bmax <= tmax) {
+			// Decode block.
+			buf, err := snappy.Decode(nil, v[8:])
+			if err != nil {
+				return fmt.Errorf("decode block: %s", err)
+			}
+
+			// Copy out any entries that aren't being overwritten.
+			for _, entry := range SplitEntries(buf) {
+				if _, ok := m[int64(btou64(entry[0:8]))]; !ok {
+					a = append(a, entry)
+				}
+			}
+
+			// Delete block in database.
+			c.Delete()
```

Comment:
just want to confirm that I'm reading this right. Because this happens in a transaction, the `Delete` and then the creation of a new block will be atomic?